cpp/src/arrow/chunk_resolver.h

```diff
@@ -97,12 +157,47 @@ struct ARROW_EXPORT ChunkResolver {
   /// \return ChunkLocation with a valid chunk_index if index is within
   ///         bounds, or with chunk_index == chunks.size() if logical index is
   ///         `>= chunked_array.length()`.
-  inline ChunkLocation ResolveWithChunkIndexHint(int64_t index,
-                                                 ChunkLocation hint) const {
+  inline ChunkLocation ResolveWithHint(int64_t index, ChunkLocation hint) const {
     assert(hint.chunk_index < static_cast<int64_t>(offsets_.size()));
     const auto chunk_index =
         ResolveChunkIndex</*StoreCachedChunk=*/false>(index, hint.chunk_index);
-    return {chunk_index, index - offsets_[chunk_index]};
+    return ChunkLocation{*this, chunk_index, index - offsets_[chunk_index]};
+  }
+
+  /// \brief Resolve `n` logical indices to chunk indices.
+  ///
+  /// \pre logical_index_vec[i] < n (for valid chunk index results)
+  /// \pre out_chunk_index_vec has space for `n` elements
+  /// \post chunk_hint in [0, chunks.size()]
+  /// \post out_chunk_index_vec[i] in [0, chunks.size()] for i in [0, n)
+  /// \post if logical_index_vec[i] >= chunked_array.length(), then
+  ///       out_chunk_index_vec[i] == chunks.size()
+  ///       and out_index_in_chunk_vec[i] is undefined (can be out-of-bounds)
+  ///
+  /// \param n The number of logical indices to resolve
+  /// \param logical_index_vec The logical indices to resolve
+  /// \param out_chunk_index_vec The output array where the chunk indices will be written
+  /// \param chunk_hint 0 or the last chunk_index produced by ResolveMany
+  /// \param out_index_in_chunk_vec If not NULLPTR, the output array where the
+  ///                               within-chunk indices will be written
+  /// \return false iff chunks.size() > std::numeric_limits<IndexType>::max()
+  template <typename IndexType>
+  [[nodiscard]] std::enable_if_t<std::is_unsigned_v<IndexType>, bool> ResolveMany(
```

Comment:
Given that there is no specialization going on, can you `static_assert` instead of `enable_it_t`? It would make the signature much more readable.

Reply:
Also, I'm not sure why the indices are supposed to be unsigned here, or why we want to support several index types.

Reply:
I'm just questioning the use of switching to unsigned. We use signed indices comprehensively in Arrow, and switching to unsigned introduces (gratuitous?) complication and inconsistency.

This is especially true if we intend to make this a public API.

Reply:
It's not a complication in the context of `Take` that already went through the trouble of validating the `indices`.

To make the API complete and compatible with the Arrow style I can add wrappers that take signed integers.

Reply:
My point is simply that validating the indices does not imply that we should change their type to unsigned.

Reply:
But this is what Wes did to simplify the dispatching of `Take` and I'm using the same reasoning to simplify the dispatching of `ResolveMany`. 

https://github.com/apache/arrow/blob/d0f3b5f3c74f67c7ca941c98bd60148e9a9e94f0/cpp/src/arrow/compute/kernels/vector_selection.cc#L465-L469

I started with signed indices and rewrote this to unsigned because I'm using this from `Take`.

I will push a commit here that exposes an API that can receive signed logical indices as well so the API is complete just like the user-exposed kernels support all integer types.